hedera-node/docs/design/services/token-service/airdrops/token-reject.md

```diff
@@ -0,0 +1,135 @@
+# Introduce TokenReject operation
+
+## Purpose
+
+As part of [HIP-904](https://hips.hedera.com/hip/hip-904), we would like to have the option of rejecting a token that has become part of an account's balance (no matter how e.g. via regular crypto transfer or airdrop).
+
+This would mean transferring the token (the whole amount if fungible) or the NFT (concrete serial number from a unique token type) back to its treasury account, without assessing custom fees. Using token reject an account could get rid of potential “ransom” tokens which have very expensive custom fees and only pay for the cheaper token reject operation.
+
+Token reject transaction can be performed only for tokens which are part of the persistent state of the account. No tokens from the pending state that is introduced with HIP-904 could be rejected. This remains only a right for the owner, who can cancel a pending airdrop.
+
+## Goals
+
+1. Define new `TokenReject` HAPI transaction
+2. Implement token reject transaction handler logic
+
+## Non-Goals
+
+- Implement token rejection in system contract functions
+
+## Architecture
+
+The implementation related to the new `TokenReject` transaction will be gated behind a feature flag.
+
+### HAPI updates
+
+Create new transaction type as defined in the HIP:
+
+```protobuf
+/**
+ * Reject undesired token(s).<br/>
+ * Transfer one or more token balances held by the requesting account to the treasury for each
+ * token type.<br/>
+ * Each transfer SHALL be one of the following
+ * - A single non-fungible/unique token.
+ * - The full balance held for a fungible/common token type.
+ *
+ * A single tokenReject transaction SHALL support a maximum of 10 transfers.
+ */
+message TokenRejectTransactionBody {
```

Comment:
Shouldn't we also have `AccountID` in the `TokenRejectTransactionBody` ?

Reply:
I think the account that is rejecting the tokens should be the payer of the transaction.
@Nana-EC is this correct?

Reply:
We should probably clarify that in the javadocs.

Reply:
Good point.
To support a different account paying for the transaction we should not assume the account and probable have an accountID field as @MrValioBg noted.

Reply:
LG